### ISSUE TITLE
[RFC] generating partial templates into an existing project

### ISSUE THREAD

**Issue by sassman** [cargo-generate/cargo-generate#291]
**As a** Developer
**I want** to generate a partial into an existing project
**So that** I can apply feature by feature to an existing project

## An example

Let's assume I've started off by a project template a while ago, the project meanwhile grew. 
Not I need the `feature` of a github workflow build pipeline in my existing project.

Let's assume further I have already a project template or a gist of my personal default github workflow pipeline that contains project specific placeholders, like some boolean flags to support building a binary and a library building out of one file. Also it may contain the crate / binary name variables. 

When I run:

```sh
cargo generate --partial my-github-handle/rust-build-pipeline .github/workflows/build.yml
```

I would trigger basically the same project as using cargo-generate the usual way. Just that:

- I can checkout a subset (here `.github/workflows/build.yml`) 
- of a given template repository (here ` my-github-handle/rust-build-pipeline`) 
- into my `$CWD` (here it would expand to `$CWD/.github/workflows/build.yml`)
- optionally I could specify a different target directory (for example `--into src/my/`)

Of course an interactive template rendering is happening as usual.

## Other Use-Cases

the above described behaviour would also enable a some more use-cases. 
For instance one could build a composable project library of `partials` that can be applied step by step. All those components could live in one template repository in purpose specific folders.

**Comment by taurr** [comment]
To be more consistent, and supporting the new `favorite` feature, maybe the syntax should be more in line with:

```shell
cargo generate --git my-github-handle/rust-build-pipeline --partial .github/workflows/build.yml
```

or with favorites:

```shell
cargo generate my-github --partial .github/workflows/build.yml
```

**However**... I am a bit conflicted as to what I think of the feature as a whole - somhow it is as if it is a simple job for a script, not really a feature?

something in the lines of:

```shell
GITPATH=$1
PATTERN=$2

TDIR=$(mktemp -d)
cargo generate --git $GITPATH --name "$TDIR/generated" --force
cp "$TDIR/generated/$PATTERN" .
```

**Comment by sassman** [comment]
It is quite usual that option arguments are not depending on a positional order (think the order cannot be enforced at all)

> However... I am a bit conflicted as to what I think of the feature as a whole - somhow it is as if it is a simple job for a script, not really a feature?

Well, one could argue that the whole feature set of of cargo-generate can be a script. 

IMO: it's all about lifting obstacles, so if I can apply certain template partials to an existing project without scripting it, it would reduce my mental load and helps me a great deal.

**Comment by taurr** [comment]
Another thought:
If we add `--init` (and thus can close e.g. #193), this could be implemented using a `--filter` option.
Allowing `--filter` 1 or more times would then allow the user to pick the needed files/folders from the generated template.

As I see it, this would allow maximum flexibility :-)

**Comment by sassman** [comment]
For skip the git init #193 there is PR #194 that still needs a little fine tuning.

Allowing the picking of single files one or multiple times sound actually pretty neat. But rather calling it `--filter` I tend to call it `--pick` (also `--partial` was not nice), but pick sounds very intuitive for me.

**Comment by taurr** [comment]
@sassman:
I like naming the option `--pick`...

**But** one thought related to my own new `--dir` option => This new `--pick` is actually just my `--dir` option, with added multiplicity and option to allow files!

Maybe we should just rename `--dir` into `--pick` now, and call it part 1 of this feature? (it would be trivial for me to do).

Even the name `--pick` resonates with #295

**Comment by taurr** [comment]
Or maybe just as an alias to `--dir`

**Comment by sassman** [comment]
Let me make a full example to avoid a little but important misunderstanding here:
```sh
cargo generate --git my-github-handle/rust-build-pipeline --pick .github/workflows/build.yml
```

would actually:

- take the git repo from `https://github.com/my-github-handle/rust-build-pipeline`
- picks from there the only the file `.github/workflows/build.yml`
- templates this file with all the configuration that is defined in the repositories `cargo-generate.toml` file
- will write this picked file into `./.github/workflows/build.yml` and would create the directories if they don't exist yet

I believe that is not quite the behaviour like the `--dir` option from #295.
Because this `--pick` option would work with `--git` or with `--dir`, so it would template things from either a git repo or a local directory into this current directory.

**Comment by taurr** [comment]
Exactly my thought - that's why I edited my previous comment a little :-)

I was just because I noted some key functionality that is actually the same internally. As stated above: `--pick` also *specifies paths relative to the template*, whereas `--dir` is *relative to the current working dir*.

**Comment by taurr** [comment]
@sassman:

> * will write this picked file into `./.github/workflows/build.yml` and would create the directories if they don't exist yet

I believe this one comment of yours is the reason for our confusion in #296 ?

I do see what you mean here - but how does this fit in with the use of e.g #194 ?

We use
```sh
cargo generate (--git <repo> | --dir <directory>) [--name <name>]
```

to generate **the full template into a new directory named _\<name\>_**.

We will use (future option not yet fully defined from #194):
```sh
cargo generate --init (--git <repo> | --dir <directory>)
```

to generate **the full template into $CWD**.

Then we have `--pick`, the purpose of which I initially thought was to pick part of the template, **and only populate the destination with the picked parts**.

Originally, I did the `--dir` part so it would work with the `--git` in a way that would allow us to have multiple templates in a single repo - picking the generated template by specifying the directory to use. As it was found confusing to be able to specify both `--git` and `--dir` at the same time, this was dropped in favor of this `--pick` option.

Whether `--pick` generates into a new directory or into $CWD like `--init` - I really don't have a strong opinion on this one. What I DO have an opinion on, is that I still miss the initial desire to use multiple templates in a single repo. Can we combine these usecases in any meaningful way?

**Comment by sassman** [comment]
> to generate the full template into $CWD.

yes

> Then we have --pick, the purpose of which I initially thought was to pick part of the template, and only populate the destination with the picked parts.

exactly, only picking the selected part, and *not* doing full fledged projected generation. Means, what you pick is what you get, into your `$CWD`, with the same name as `picked`. 

So when `--pick .github/workflows/build.yml` it will take that one from the template (git or dir) and put it "rendered" just into there `$CWD`. So `$CWD/.github/workflows/build.yml` would be the result.

That means you could generate what ever artifacts in what ever folder, it could run everywhere, not limited to a create or a git repository.

**Comment by taurr** [comment]
Come to think about it, `--git` and `--dir` both specifies the **source** of the template.

`--init` alters the **destination**.

`--pick` specifies **what** to take from the template (And in this case I *still* think it should create a new folder unless `--init` is specified).

What I still miss, is the **source** option to specify a subdir within a repository - which will allow us to have multiple templates in a single repo. I originally used `--dir` for this, and I still think this would be ok :-/

**Comment by taurr** [comment]
This said, my opinion on the destination of `--pick` is weak, and I will make it so it goes to $CWD as that also has it's merrits.

However, I *strongly* feel the need for this part:
> What I still miss, is the source option to specify a subdir within a repository - which will allow us to have multiple templates in a single repo. I originally used --dir for this, and I still think this would be ok :-/

Please allow us to combine `--git` and `--dir` - or come up with another option name for specifying the **source* within a git repository.

**Comment by sassman** [comment]
> What I still miss, is the source option to specify a subdir within a repository - which will allow us to have multiple templates in a single repo. I originally used --dir for this, and I still think this would be ok :-/

Ok, I'm not sure If I do see a contradiction here. Above we have discussed the `--pick` with a file. That is fine. Let's have a look how that would work with a directory. 

## Given
- a git repo or directory X that contains 2 folders like below:
```plain
.
├── cli-template
│   ├── Cargo.toml
│   ├── LICENSE
│   ├── README.md
│   ├── cargo-generate.toml
│   └── src
│       └── main.rs
└── rust-bin-template
    ├── Cargo.toml
    ├── LICENSE
    ├── README.md
    └── src
        └── main.rs
```
- also given I'm in a certain project Y that I want to add the whole subtree `X/rust-bin-template` into

## Then
I would call `cargo generate --dir /path/to/X --pick rust-bin-template` and in my current directory I would get a templated version of `rust-bin-template`, just not renamed and verified to be a valid crate. It would be a 1-1 copy of the full folder structure and everything got templated.

Would that not cover the case you are seeking?

**Comment by taurr** [comment]
It would, but `--pick` would not cause cargo-generate.toml to be picked up in the sub-folder.

It only alters *what* gets copied to the destination, and `cargo-generate` only looks for its files in the root of the *source*

**Comment by sassman** [comment]
> It only alters what gets copied to the destination, and cargo-generate only looks for its files in the root of the source

That is a good thing to discuss. 
IMO `--pick` should still pick up the `cargo-generate.toml` in the given repo / dir. So that this part is also considered.

For subfolders containing this file, I have not yet a clear opinion how to handle it. But maybe a suggestion: if the argument of `--pick` is a folder that contains a `cargo-generate.toml` it would be preferred over the root `cargo-generate.toml`.

Thoughts?

**Comment by taurr** [comment]
@sassman 

Not UML, but grafically, this is aproximately how it works atm: ![asd](http://www.plantuml.com/plantuml/svg/dL6zZjim3Dxv5ApHrLV4umtN0t9j7w01EZGT18bsX9D5GwBH2vfxzz9AP1ZDLWs9J7q_v4TupcAIp6F05ve2dqUA0iOA9vmp9J235h2n9WtSwfsI4MhGPBG3sC2pmynpiWGz1qS9cw-9yWajJsVa7eN6ALIAZDCA-_8scUZ8tI4wWCKPsoUrm4TishRvENYvE9z-LJDzUNXusdor2bUouh-HO5zwgCeLTlrJOeNk0qT2JgLRq5gLwVJnsk0g0dqo8sbAMRvJ82lu1r3FYTxCm14R7xtWvOAJjw-Vop0yYUTeJwy5M0V_kF_nSWtmjzjTDJRFJlLzD0jnmRWikyLeSqmItnlrR-ej7-F0saYY3qLuhUzMuZkil_UDg4cd9lzXqDrauHrAQXhVPkhrLkfbb4UCV5zVuZvJ1h0S9N70PgyBALO7s2Ikw0CS2kkWb0LqZ6sx98wz3vG_EP-qsJxnU5lwCGflGrMwxdyJqR-53fufkdaCy1S0)

I feel like `--pick` is getting too much responsibility if it is to mess with the location of `cargo-generate.toml` too. Also, what if the user specifies multiple `--pick`'s ?

IMHO `--pick` is used to pick apart the template, and only grab what you need, *after* it has been altered etc.

I would *much* rather have a new option for specifying the sub-location of the actual template within the source, than to have `--pick` involved in the mess it will become to select the right `cargo-generate.toml` (and maybe values files and other stuff, needed to properly expand the template).

## Proposal

* `--git` selects the source repository (unchanged).
* `--dir` selects a specific *local* template directory, repository or not (unchanged).
* `--pick` can be specified *multiple times*, and specifies *what* gets copied to the final destination (and alters this to always be `$CWD` - just as you like)
* `--template-dir`? Used together with `--git` to specify a directory within the repository to be used as the template.

This would also allow us to e.g. specify an option that looks for multiple locations of `cargo-generate.toml` inside a repo, then lets the user select the template to expand interactively. Heck, I would even make this behavior the default, as it would not change anything for existing template repositories that only contain 1 template in the root.

Obviously this option is only to be used with `--git` as it is rather pointless with `--dir`.

**Comment by taurr** [comment]
This diagram might better explain how I saw it:

![diagram](http://www.plantuml.com/plantuml/svg/ZLB1ZjCm4BtxAqOgXzPIDkBQBjMYHUB2WGCIPsyoJKTnFCPsGD66_NOSfuaS5X0-MFRupRmtRomUV50kz9s6zrMWxnG6aIA2Bh1_AOg60aZfkNSLIdc0S4K3-m7zGK1Sn-czqcmmGm6PmF4UEYlbc70sbgH0RA0e0dPMgu15JGvsLcbi7AA1vuISURSWAUF-cZsjtITghk6Zm-4evP8LDRnR19Itx3AbeFO8UyD_B73_T5ugt9TFvwp6YZOS4DoK8I1RZzph6ZHp2nTsK2dNSD6WGHThtGNkz0QkJ3sP55ydqr4Fe7mBFwvKNI4mz1utS3AneaSulmNoaQ0tzTqAG5CdLU9seKiw_HHhgv0R6xturq3JK2ksGpxFV9h_x_lV_BPasYCwn-wq1blIEcjWtiM6mwXkhX-RcCumZhc7fJXQnnLw3xcUKIm_sLBLmhRXxSzCWDrdfV5BxxxrJ1xXULNeqA8AYpShGpbbpHQzUVpwePHo3fn-Jup3rT69ZXt2eeRjHEmFG7vLb4yktLzJliYSC1cJOgURRpwm5G_nt7TQY5y0)

**Comment by taurr** [comment]
or even like this:
![diagram](http://www.plantuml.com/plantuml/svg/dLBDYjim4BxhAGQbXyIGfhrwAIoqq4kXFHHw5iua6I9hL4dk4jRPPw-ahcN7-D2D3uuy6Nq_Cz-Z3yg5hjKY7mGTOFsotHufG5Lvxbo3LRM1S483wmlwZO3urEd_HhD1a7BianB81CwrqDfKw7E_ZT0KY0q4RArM0RTxSh2oIkFHOSH-pds9NSh8VzDMLV6TmEOwyhFwHiTJ-EhmKYVUuMPMyscKiYiektmLjKTO6tupK04mUphaqrKKCBYraTMFurasK3TiB_CvpgPubyXbbvRCkKRds3sKufcqdfZwznRZrkHtctwLbg-BgZKtaJOIDieTkUphVU1MHoa7qkZxCemFSE3ExEVs_zVKKCYk1bi9z-CIxXt0CHmPw7PmuuQKFwTaerDDe3-eBm_Y3Wfb_1EwUOAkOWPXL5kkfn2bRna_-twgOITgwxX1xs7AsWjHqcUfESDIwiR5vfRL3wNnP-T-TqmUuRbuSsXHXKb0YvbqRypfky-_laH_Gx6V0poTI2EiMtO8HLcon3bMVWFa9_AkuYQsI_mlrOmzZssezwy9ze6jU8pdhjL_0G00)

**Comment by taurr** [comment]
In your specific case

```sh
cargo generate --dir /path/to/X --template-dir rust-bin-template
```

would do _almost_ what you want. namely generate

```
.
|_ rust-bin-template
  |_ Cargo.toml
    |_ LICENSE
    |_ README.md
    |_ src
      |_ main.rs
```

If you wanted it in the current dir, I guess we would need #193... that would be something along the lines of

```sh
cargo generate --dir /path/to/X --template-dir rust-bin-template --init
```

whereas, if I **only** wanted e.g. the license file, I could do (and here I grant it would be stupid to create the project dir):

```sh
cargo generate --dir /path/to/X --template-dir rust-bin-template --pick LICENSE
```

which would generate

```
.
|_ LICENSE
```

**Comment by sassman** [comment]
That was very helpful, the `--init` option does make sense now and very nice drawing btw. 

However, the the `--dir` and `--git` should both only take care of selecting the source that will be processed. IMO that is the stage of collecting the sources, and should behave the same, for sources from git and sources from dir. Means at this point also a config file in the source dir (`--dir` or `--git`) would have been processed. 

After that picking a file or a folder would be the next stage. That could even be applied multiple times. I understand your suggestion of differntiate between `--template-dir` and `--pick` but I'd rather not like to go down this route of even more parameter complexity route. 

I'd really like to KISS that simple that there is only one parameter that might be applied multiple times, that is responsible to sub-select something (folder or file) from a given source tree. Like the `--pick` was described. 

I believe that the mental load of `--dir` with `--template-dir` and `--pick` all-together might cause serious brain damage to our users and should be simplified to:

- select the source: `--git` or `--dir`
- select what to pick: `--pick file` or `--pick sub-dir` or `--pick file1  --pick folder2` 
- decide if it should go into an existing project or not 
  - a new project: `--init`
  - an existing project: no further argument

those 3 layers of decisions are straight forward and easy to imagine. A route of `--git` and `--pick` would behave exactly the same as `--dir` and `--pick` no additional mental load or side thoughts needed.

**Comment by taurr** [comment]
@sassman

Please note that I did/do NOT intend to use `--template-dir` together with `--dir` as the user could just use `--dir` alone! Using these 2 together does not make any sense.

On the behavior of `--pick` - I must admit that I do not see how these can easily be merged without conflicting and confusing behavior.

## Common usecases

### Scenario 1.1 (current)

#### Given

```
my-local-dir
|_ local-crate
  |_ .git
  | |_ ...
  |_ Cargo.toml
  |_ src
    |_ lib.rs
```

##### and

```
my-template
|_ Cargo-generate.toml
|_ Cargo.toml
|_ src
  |_ lib.rs
```

#### Then

```sh
cargo generate --git <my-template>
```

* Asks for project name (default is `my-template`) or uses `--name <project-name>`.

Generates:

```
my-local-dir
|_ local-crate
| |_ .git
| | |_ ...
| |_ Cargo.toml
| |_ src
|   |_ lib.rs
|_ <project-name>
  |_ .git
  | |_ ...
  |_ Cargo.toml
  |_ src
    |_ lib.rs

```

### Scenario 1.2 (current)

#### Given

```
my-local-dir
|_ .git
| |_ ....
|_ Cargo.toml
|_ src
  |_ lib.rs
```

##### and

```
my-template
|_ Cargo-generate.toml
|_ Cargo.toml
|_ src
  |_ lib.rs
```

#### Then

```sh
cargo generate --git <my-template>
```

* Asks for project name (default is `my-template`), or uses `--name <project-name>`.
* Does not initialize a new git repository.

Generates:

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
| |_ lib.rs
|_ <project-name>
  |_ Cargo.toml
  |_ src
    |_ lib.rs
```

⚠️ In my mind, we should at least give a warning here, as the local `Cargo.toml` is not a workspace!

### Scenario 1.3 (new)

#### Given

```
my-local-dir
|_ local-crate
| |_ .git
| | |_ ...
  |_ Cargo.toml
  |_ src
    |_ lib.rs
```

##### and

```
my-template
|_ Cargo-generate.toml
|_ Cargo.toml
|_ src
  |_ lib.rs
```

#### Then

```sh
cargo generate --git <my-template> --pick src
```

* Asks for project name (default is `my-template`), or uses `--name <project-name>`.
* ⚠️ Does not initialize a new repository - if it did, it would conflict with the one inside `local-crate`, and `src` is not a template dir... so whats the rule here!

Generates:

```
my-local-dir
|_ local-crate
| |_ .git
| | |_ ...
| |_ Cargo.toml
| |_ src
|   |_ lib.rs
|_ src
  |_ lib.rs
```

### Scenario 1.4

#### Given

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
  |_ lib.rs
```

##### and

```
my-template
|_ Cargo-generate.toml
|_ Cargo.toml
|_ src
  |_ lib.rs
```

#### Then

```sh
cargo generate --git <my-template> --pick src
```

* Asks for project name (default is `my-template`), or uses `--name <project-name>`
* *Could have read `project-name` from the local `Cargo.toml`.*
* Does not initialize a new repository
* ⚠️ It **overwrites** the local `src` directory!

Generates
```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
  |_ lib.rs    <-- new from template!!!
```

Need to define whether to abort or warn in this case.

## Usescases with "merged" behavior

### Scenario 2.1

#### Given

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
  |_ lib.rs
```

##### and

```
my-template
|_ template1
| |_ Cargo-generate.toml
| |_ README.md
|_ template2
  |_ Cargo-generate.toml
  |_ Cargo.toml
  |_ README.md
  |_ src
    |_ lib.rs
```

#### Then

```sh
cargo generate --git <my-template> --pick template2
```

* Picks up `Cargo-generate.toml` from inside `template2`
* Asks for project name (default is **???**), or uses `--name <project-name>`
* *Could have read `project-name` from the local `Cargo.toml`.*
* ⚠️ It generates the template files as usual, but because of `--pick` it doesn't create the project dir

Generates:

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
| |_ lib.rs
|_ template2    <-- normally I would expect this to be named <project-name>, like scenario 1.1
  |_ Cargo.toml
  |_ README.md
  |_ src
    |_ lib.rs
```

### Scenario 2.2

Multiple `--pick` options create isses with template selection.

Preconditions like scenario 2.1

#### Then
```sh
cargo generate --git <my-template> --pick template2 --pick template1/README.md
```

* Picks up `Cargo-generate.toml` from inside `template2` due to the first `--pick`
* Asks for project name (default is **???**), or uses `--name <project-name>`.
* *Could have read `project-name` from the local `Cargo.toml`.*
* ⚠️ Generates everything in the template as if it belongs to the same template - even though we've picked files from 2 different templates!
    * Alternative, we have to ask for the `project-name` again, and regenerate `README.md` in that context! 

Generates:

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
| |_ lib.rs
|_ template2    <-- normally I would expect this to be named <project-name>, like scenario 1.1
| |_ Cargo.toml
  |_ README.md
| |_ src
|   |_ lib.rs
|_ README.md    <-- this file doesn't actually belong to the same template!
```

Notice there's a lot of issues here with `--pick` from different places, spanning possible multiple templates.

## Usescases with extra option

### Scenario 3.1

Preconditions like scenario 2.1

#### Then

```sh
cargo generate --git <my-template> --pick template2
```

* Would *not* pick up `Cargo-generate.toml` from inside `template2`. `--pick` *only* specifies what to take from the template!

Whereas

```sh
cargo generate --git <my-template> --templatedir template2
```

* Picks up `Cargo-generate.toml` from inside `template2` - No ambiguity.
* Asks for project name (default is **template2**), or uses `--name <project-name>`
* Renders only the files inside the `template2` dir - and in that context.

Generates:

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
| |_ lib.rs
|_ <project-name>
  |_ Cargo.toml
  |_ README.md
  |_ src
    |_ lib.rs
```

### Scenario 2.4

Multiple `--pick` options does **not** create isses with template selection.

Preconditions like scenario 2.1

#### Then

```sh
cargo generate --git <my-template> --templatedir template2 --pick template2 --pick template1/README.md
```

Will generate error, as the 2 `--pick` options specify files that do not exist in the context of `template2`.

```sh
cargo generate --git <my-template> --templatedir template2 --pick src --pick README.md
```

* Picks up `Cargo-generate.toml` from inside `template2` - No ambiguity.
* Asks for project name (default is **template2**), or uses `--name <project-name>`.
* Renders only the files inside the `template2` dir - and in that context.
* Does not allow picking files from different templates

##### Generates (when `--pick` behaves somewhat like `--init`)

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
| |_ lib.rs    <-- overwritten from template2!
|_ README.md
```

##### Generates (when `--pick` does not behave  like `--init`)

```
my-local-dir
|_ .git
| |_ ...
|_ Cargo.toml
|_ src
| |_ lib.rs
|_ <project-name>
  |_ README.md
  |_ src
    |_ lib.rs
```

**Comment by sassman** [comment]
@taurr thanks for that summary, I'm still a bit overwhelmed and need to process that all.

**Comment by taurr** [comment]
Well, I know what I prefer and as I said; I don't even see how to get it working properly otherwise.

But I'm really not the deciding factor here - what do other users think?

**Comment by sassman** [comment]
is solved by #379
